### PR TITLE
Preferences Window

### LINKED ISSUE #140: Settings page
We'll need a settings page that shows two main areas:

- Program settings
- Archive settings

There are many things (like timeouts) that need to be configured.

### LINKED ISSUE #565: [feature request] option to (auto)start (minimized) with OS
Such options would be handy/welcome now that the tray icon been implemented.

tray option

### LINKED ISSUE #587: Auto lock unlocked archives after time
Based on #141 - Auto lock the archive after **x** minutes of inactivity. The following points should be covered:

 * Lock all currently unlocked archives
 * Listen for user-cues in terms of activity
   * Reset lock timer on mouse move
   * Reset lock timer on focus change
   * Reset lock timer on tray icon menu open
 * Disable the timer while any item is being edited

### LINKED ISSUE #702: Global hotkey
Hi, thanks for this awesome app.

Can we add a global hotkey for maximizing/minimizing buttercup?
It would be also nice to have a preferences dialog where we can customize these things.

### PR DESCRIPTION
**Pull request by sallar**
Fixes #140
Fixes #702
Fixes #587
Fixes #565

### CODE REVIEW

**Comment by ph1p** [comment]
I have removed all empty translations https://github.com/buttercup/buttercup-desktop/pull/883

File: locales/de/base.json [hunk 1]
```diff
@@ -117,7 +117,10 @@
     "name": "Name",
     "size": "Größe",
     "date": "Datum",
-    "new-archive": "Neues Archiv"
+    "new-archive": "Neues Archiv",
+    "connect-to-googledrive": "",
```
Comment (perry-mitchell) [inline locales/de/base.json]: These can be removed by https://github.com/buttercup/buttercup-desktop/pull/883

File: src/renderer/system/archives.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  lockAllArchives() {
+    const archives = getAllArchives(store.getState());
+    if (archives) {
+      archives.forEach(
+        ({ id, status }) =>
+          status === 'unlocked' ? store.dispatch(lockArchive(id)) : ''
+      );
+    }
+  },
+  lockArchiveTimer() {
```
Comment (perry-mitchell) [inline src/renderer/system/archives.js]: This should be called when there's activity, too. It shouldn't lock while in focus if there's activity going on.

File: src/renderer/system/archives.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      lockArchiveOnFocusout,
+      isButtercupFocused
+    } = state.settings;
+
+    if (
+      (autolockSeconds === '0' &&
+        (lockArchiveOnFocusout && !isButtercupFocused)) ||
+      (autolockSeconds !== '0' &&
+        ((lockArchiveOnFocusout && !isButtercupFocused) ||
+          !lockArchiveOnFocusout))
```
Comment (perry-mitchell) [inline src/renderer/system/archives.js]: This logic should check to see if there is any editing ongoing. If there is, this should fail so no timer is started.

Comment (ph1p) [reply src/renderer/system/archives.js]: I'll try to solve all suggestions in #883

Comment (ph1p) [reply src/renderer/system/archives.js]: Added some checks

File: package.json [hunk 4]
```diff
@@ -143,6 +143,7 @@
   },
   "homepage": "https://buttercup.pw",
   "dependencies": {
+    "eva-icons": "^1.1.1",
```
Comment (perry-mitchell) [inline package.json]: @sallar This alright?

Comment (sallar) [reply package.json]: @perry-mitchell I will move it to dev deps myself

**Comment by sallar** [comment]
@ph1p I updated the description of this PR to auto-close the related issues. Can you check to see if this is everything that is related?

**Comment by sallar** [comment]
@perry-mitchell can I get another review?

**Comment by ph1p** [comment]
@sallar Sure!

**Comment by ph1p** [comment]
I've looked through everything and found no other issues, just a few that would maybe fit into the settings (:

#806
#886

**Comment by sallar** [comment]
Thanks @ph1p, since we havent implemented those features yet, I dont think we can close them. So I'm gonna merge this as is. Thank you both

**Comment by ph1p** [comment]
> Thanks @ph1p, since we havent implemented those features yet, I dont think we can close them. So I'm gonna merge this as is. Thank you both

I didn't mean to imply that we could close them. Just that they might be implemented in the future. Sorry for the confusion.

**Comment by perry-mitchell** [comment]
Woot! To master we go!

image

Really sorry this took so long @ph1p - But this is epic work. Will be excellent to see it in action!

**Pull request merged**